File: solutions/53.Maximum_Subarray/53.Maximum_Subarray.py
from ast import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = nums[0]
        current_sum = nums[0]

        for i in range(1, len(nums)):
            num = nums[i]
            current_sum = max(current_sum + num, num)
            max_sum = max(current_sum, max_sum)

        return max_sum

File: solutions/medium/53.Maximum_Subarray/53.Maximum_Subarray.ts
function maxSubArray(nums: number[]): number {
  let currentSum = nums[0];
  let maxSum = nums[0];
  for(let i = 1; i<nums.length; i++){
      let num = nums[i]
      currentSum = Math.max(currentSum + num, num)
      maxSum = Math.max(currentSum, maxSum)
  }
  return maxSum;
};

//or 

function maxSubArraySecond(nums: number[]): number {
  let maxCurrent = nums[0];
  let maxGlobal = nums[0];
  for(let i = 1; i<nums.length; i++){
      maxCurrent = Math.max(nums[i] , maxCurrent + nums[i])
      
      if(maxCurrent > maxGlobal){
          maxGlobal = maxCurrent;
      }
  }
  return maxGlobal;
};